c_src/scenic/ops/cairo_script_ops.c

```diff
@@ -0,0 +1,663 @@
+  case TEXT_ALIGN_CENTER:
+    align_offset = -(text_extents.width / 2);
+    break;
+  case TEXT_ALIGN_RIGHT:
+    align_offset = -(text_extents.width);
+    break;
+  }
+
+  float base_offset;
+  switch (p_ctx->text_base) {
+  case TEXT_BASE_TOP:
+    base_offset = font_extents.ascent;
+    break;
+  case TEXT_BASE_MIDDLE:
+    base_offset = -((font_extents.descent - font_extents.ascent) / 2);
+    break;
+  case TEXT_BASE_ALPHABETIC:
+    base_offset = 0;
+    break;
+  case TEXT_BASE_BOTTOM:
+    base_offset = -(font_extents.descent);
+    break;
+  }
+
+  cairo_translate(p_ctx->cr, align_offset, base_offset);
+
+  if (status == CAIRO_STATUS_SUCCESS) {
+    int glyph_index = 0;
+    int byte_index = 0;
+    for (int i = 0; i < cluster_count; i++) {
+      cairo_text_cluster_t* cluster = &clusters[i];
+      cairo_glyph_t* cluster_glyphs = &glyphs[glyph_index];
+
+      cairo_scaled_font_glyph_extents(scaled_font, cluster_glyphs, cluster->num_glyphs, &text_extents);
+      cairo_show_glyphs(p_ctx->cr, cluster_glyphs, cluster->num_glyphs);
+
+      glyph_index += cluster->num_glyphs;
+      byte_index += cluster->num_bytes;
+    }
+  } else {
+    log_error("%s: cairo_scaled_font_text_to_glyphs: error %d", __func__, status);
+  }
+
+  p_ctx->text_align = TEXT_ALIGN_LEFT;
+  p_ctx->text_base = TEXT_BASE_ALPHABETIC;
+}
+
+#warning "cairo TODO: script_ops_draw_sprites"
+
+void script_ops_begin_path(void* v_ctx)
+{
+  if (g_opts.debug_mode) {
+    log_script_ops_begin_path(log_prefix, __func__, log_level_info);
+  }
+
+  CAIROcontext_t* p_ctx = (CAIROcontext_t*)v_ctx;
+  cairo_new_path(p_ctx->cr);
+}
+
+void script_ops_close_path(void* v_ctx)
+{
+  if (g_opts.debug_mode) {
+    log_script_ops_close_path(log_prefix, __func__, log_level_info);
+  }
+
+  CAIROcontext_t* p_ctx = (CAIROcontext_t*)v_ctx;
+  cairo_close_path(p_ctx->cr);
+}
+
+void script_ops_fill_path(void* v_ctx)
+{
+  if (g_opts.debug_mode) {
+    log_script_ops_fill_path(log_prefix, __func__, log_level_info);
+  }
+
+  CAIROcontext_t* p_ctx = (CAIROcontext_t*)v_ctx;
+  cairo_fill(p_ctx->cr);
+}
+
+void script_ops_stroke_path(void* v_ctx)
+{
+  if (g_opts.debug_mode) {
+    log_script_ops_stroke_path(log_prefix, __func__, log_level_info);
+  }
+
+  CAIROcontext_t* p_ctx = (CAIROcontext_t*)v_ctx;
+  cairo_stroke(p_ctx->cr);
+}
+
+void script_ops_move_to(void* v_ctx,
+                        coordinates_t a)
+{
+  if (g_opts.debug_mode) {
+    log_script_ops_move_to(log_prefix, __func__, log_level_info,
+                           a);
+  }
+
+  CAIROcontext_t* p_ctx = (CAIROcontext_t*)v_ctx;
+  cairo_move_to(p_ctx->cr, a.x, a.y);
+}
+
+void script_ops_line_to(void* v_ctx,
+                        coordinates_t a)
+{
+  if (g_opts.debug_mode) {
+    log_script_ops_line_to(log_prefix, __func__, log_level_info,
+                           a);
+  }
+
+  CAIROcontext_t* p_ctx = (CAIROcontext_t*)v_ctx;
+  cairo_line_to(p_ctx->cr, a.x, a.y);
+}
+
+#warning "cairo TODO: script_ops_arc_to"
```

Comment:
I would probably change the TODO to unimplemented. And add a ticket to our backlog to implement those.

```suggestion
#warning "cairo: script_ops_arc_to unimplemented"
```

Reply:
Done